File: src/gui/KeyBinder.java
package gui;

import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.util.Hashtable;

public abstract class KeyBinder implements KeyListener
{
    private Hashtable<Integer, Boolean> keyMap;

    private int[] keyCodes;

    public KeyBinder(final int... keyCodes)
    {
        this.keyMap = new Hashtable<>();
        this.keyCodes = keyCodes;
    }

    @Override
    public void keyTyped(final KeyEvent e) { }

    @Override
    public void keyPressed(final KeyEvent e)
    {
        getKeyMap().put(e.getKeyCode(), true);

        if (getKeysDown())
        {
            onKeysDown();
        }
    }

    @Override
    public void keyReleased(final KeyEvent e)
    {
        getKeyMap().remove(e.getKeyCode(), true);
    }

    private Hashtable<Integer, Boolean> getKeyMap()
    {
        return this.keyMap;
    }

    public abstract void onKeysDown();

    public boolean getKeysDown()
    {
        for (final int key : this.keyCodes)
        {
            if (getKeyMap().containsKey(key))
            {
                if (!getKeyMap().get(key))
                {
                    return false;
                }
            } else {
                return false;
            }
        }

        getKeyMap().clear();
        return true;
    }
}

File: src/main/java/gui/KeyBinder.scala
package gui

import java.awt.event.{KeyEvent, KeyListener}
import java.util

abstract class KeyBinder(val keyCodes: Int*) extends KeyListener {
    private val keyMap: util.Hashtable[Int, Boolean] = new util.Hashtable[Int, Boolean]

    override def keyTyped(e: KeyEvent): Unit = {}

    override def keyPressed(e: KeyEvent): Unit = {
        keyMap.put(e.getKeyCode, true)
        if (getKeysDown) {
            onKeysDown()
        }
    }

    override def keyReleased(e: KeyEvent): Unit =
        keyMap.remove(e.getKeyCode, true)

    private def getKeysDown: Boolean = {
        for (key <- this.keyCodes) {
            if (keyMap.containsKey(key)) {
                if (!keyMap.get(key))
                    return false
            }
            else
                return false
        }
        keyMap.clear()

        true
    }

    def onKeysDown(): Unit
}
